File: sdks/python-admin/aegis_db_admin/vault.py
"""
Aegis-DB Admin SDK - Vault Service

Secrets management, transit encryption, and audit logging.

@version 1.0.0
@author AutomataNexus Development Team
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TYPE_CHECKING

from .types import VaultStatus, Secret, TransitKey, TransitResult, AuditEntry

if TYPE_CHECKING:
    from .client import AegisAdmin


class VaultService:
    """Manage the integrated secrets vault."""

    def __init__(self, admin: AegisAdmin) -> None:
        self._admin = admin

    # =========================================================================
    # Status / Seal
    # =========================================================================

    async def get_status(self) -> VaultStatus:
        """Get vault status.

        Returns:
            VaultStatus object.
        """
        data = await self._admin._request("GET", "/api/v1/vault/status")
        return VaultStatus(
            sealed=data.get("sealed", True),
            initialized=data.get("initialized", False),
            version=data.get("version"),
            _raw=data,
        )

    async def seal(self) -> Dict[str, Any]:
        """Seal the vault.

        Returns:
            Server response dict.
        """
        return await self._admin._request("POST", "/api/v1/vault/seal")

    async def unseal(self, key: Optional[str] = None) -> Dict[str, Any]:
        """Unseal the vault.

        Args:
            key: Unseal key. May be omitted if the vault uses auto-unseal.

        Returns:
            Server response dict.
        """
        payload: Dict[str, Any] = {}
        if key is not None:
            payload["key"] = key
        return await self._admin._request("POST", "/api/v1/vault/unseal", payload)

    # =========================================================================
    # Secrets
    # =========================================================================

    async def list_secrets(self) -> List[Secret]:
        """List all secrets (keys only, values redacted).

        Returns:
            List of Secret objects (values may be None).
        """
        data = await self._admin._request("GET", "/api/v1/vault/secrets")
        secrets_list = data if isinstance(data, list) else data.get("secrets", [])
        return [_parse_secret(s) for s in secrets_list]

    async def get_secret(self, key: str) -> Secret:
        """Get a secret by key.

        Args:
            key: The secret key.

        Returns:
            Secret object with value.
        """
        data = await self._admin._request("GET", f"/api/v1/vault/secrets/{key}")
        return _parse_secret(data)

    async def set_secret(self, key: str, value: Any) -> Secret:
        """Create or update a secret.

        Args:
            key: The secret key.
            value: The secret value.

        Returns:
            The stored Secret.
        """
        payload: Dict[str, Any] = {"value": value}
        data = await self._admin._request(
            "PUT", f"/api/v1/vault/secrets/{key}", payload
        )
        return _parse_secret(data)

    async def delete_secret(self, key: str) -> Dict[str, Any]:
        """Delete a secret.

        Args:
            key: The secret key.

        Returns:
            Server response dict.
        """
        return await self._admin._request("DELETE", f"/api/v1/vault/secrets/{key}")

    # =========================================================================
    # Transit encryption
    # =========================================================================

    async def transit_encrypt(
        self,
        plaintext: str,
        *,
        key_name: Optional[str] = None,
    ) -> TransitResult:
        """Encrypt data using transit encryption.

        Args:
            plaintext: The plaintext data to encrypt.
            key_name: Optional transit key name to use.

        Returns:
            TransitResult with ciphertext.
        """
        payload: Dict[str, Any] = {"plaintext": plaintext}
        if key_name is not None:
            payload["key_name"] = key_name

        data = await self._admin._request(
            "POST", "/api/v1/vault/transit/encrypt", payload
        )
        return TransitResult(
            ciphertext=data.get("ciphertext"),
            plaintext=None,
            key_version=data.get("key_version"),
            _raw=data,
        )

    async def transit_decrypt(
        self,
        ciphertext: str,
        *,
        key_name: Optional[str] = None,
    ) -> TransitResult:
        """Decrypt data using transit encryption.

        Args:
            ciphertext: The ciphertext to decrypt.
            key_name: Optional transit key name to use.

        Returns:
            TransitResult with plaintext.
        """
        payload: Dict[str, Any] = {"ciphertext": ciphertext}
        if key_name is not None:
            payload["key_name"] = key_name

        data = await self._admin._request(
            "POST", "/api/v1/vault/transit/decrypt", payload
        )
        return TransitResult(
            ciphertext=None,
            plaintext=data.get("plaintext"),
            key_version=data.get("key_version"),
            _raw=data,
        )

    async def create_transit_key(
        self,
        name: str,
        *,
        algorithm: Optional[str] = None,
    ) -> TransitKey:
        """Create a new transit encryption key.

        Args:
            name: Key name.
            algorithm: Encryption algorithm (e.g. 'aes256-gcm').

        Returns:
            The created TransitKey.
        """
        payload: Dict[str, Any] = {"name": name}
        if algorithm is not None:
            payload["algorithm"] = algorithm

        data = await self._admin._request(
            "POST", "/api/v1/vault/transit/keys", payload
        )
        return _parse_transit_key(data)

    async def list_transit_keys(self) -> List[TransitKey]:
        """List transit encryption keys.

        Returns:
            List of TransitKey objects.
        """
        data = await self._admin._request("GET", "/api/v1/vault/transit/keys")
        keys_list = data if isinstance(data, list) else data.get("keys", [])
        return [_parse_transit_key(k) for k in keys_list]

    # =========================================================================
    # Audit
    # =========================================================================

    async def get_audit_log(self) -> List[AuditEntry]:
        """Get the vault audit log.

        Returns:
            List of AuditEntry objects.
        """
        data = await self._admin._request("GET", "/api/v1/vault/audit")
        entries = data if isinstance(data, list) else data.get("entries", [])
        return [
            AuditEntry(
                id=e.get("id"),
                operation=e.get("operation"),
                path=e.get("path"),
                user=e.get("user"),
                timestamp=e.get("timestamp"),
                status=e.get("status"),
                _raw=e,
            )
            for e in entries
        ]


# =============================================================================
# Helpers
# =============================================================================

def _parse_secret(data: Dict[str, Any]) -> Secret:
    return Secret(
        key=data.get("key", ""),
        value=data.get("value"),
        version=data.get("version"),
        created_at=data.get("created_at"),
        updated_at=data.get("updated_at"),
        _raw=data,
    )


def _parse_transit_key(data: Dict[str, Any]) -> TransitKey:
    return TransitKey(
        name=data.get("name", ""),
        algorithm=data.get("algorithm"),
        version=data.get("version"),
        created_at=data.get("created_at"),
        _raw=data,
    )

File: sdks/javascript-admin/src/vault.ts
/**
 * Vault service for secrets management and transit encryption.
 */

import type { RequestFn } from './index';

// ============================================================================
// Types
// ============================================================================

export interface VaultStatus {
  initialized: boolean;
  sealed: boolean;
  version?: string;
  cluster_name?: string;
  secret_count?: number;
}

export interface Secret {
  key: string;
  value: string;
  version: number;
  created_at: string;
  updated_at: string;
  metadata?: Record<string, string>;
}

export interface SecretListResponse {
  secrets: SecretEntry[];
}

export interface SecretEntry {
  key: string;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface PutSecretRequest {
  value: string;
  metadata?: Record<string, string>;
}

export interface UnsealRequest {
  key: string;
}

export interface UnsealResponse {
  sealed: boolean;
  progress: number;
  threshold: number;
}

export interface TransitEncryptRequest {
  plaintext: string;
  key_name: string;
}

export interface TransitEncryptResponse {
  ciphertext: string;
  key_version: number;
}

export interface TransitDecryptRequest {
  ciphertext: string;
  key_name: string;
}

export interface TransitDecryptResponse {
  plaintext: string;
}

export interface TransitKeyRequest {
  name: string;
  type?: string;
  exportable?: boolean;
}

export interface TransitKey {
  name: string;
  type: string;
  latest_version: number;
  min_decryption_version: number;
  exportable: boolean;
  created_at: string;
}

export interface AuditEntry {
  id: string;
  timestamp: string;
  operation: string;
  path: string;
  user?: string;
  source_ip?: string;
  success: boolean;
}

export interface AuditLogResponse {
  entries: AuditEntry[];
}

// ============================================================================
// Service
// ============================================================================

export class VaultService {
  constructor(private request: RequestFn) {}

  /** Get vault status (sealed/unsealed, initialized). */
  async getStatus(): Promise<VaultStatus> {
    return this.request<VaultStatus>('GET', '/api/v1/vault/status');
  }

  /** Seal the vault. */
  async seal(): Promise<{ message: string }> {
    return this.request<{ message: string }>('POST', '/api/v1/vault/seal');
  }

  /** Unseal the vault with a key share. */
  async unseal(key: string): Promise<UnsealResponse> {
    return this.request<UnsealResponse>('POST', '/api/v1/vault/unseal', { key });
  }

  /** List all secrets (keys only, no values). */
  async listSecrets(): Promise<SecretListResponse> {
    return this.request<SecretListResponse>('GET', '/api/v1/vault/secrets');
  }

  /** Get a secret by key. */
  async getSecret(key: string): Promise<Secret> {
    return this.request<Secret>('GET', `/api/v1/vault/secrets/${encodeURIComponent(key)}`);
  }

  /** Create or update a secret. */
  async putSecret(key: string, value: string, metadata?: Record<string, string>): Promise<Secret> {
    return this.request<Secret>('PUT', `/api/v1/vault/secrets/${encodeURIComponent(key)}`, {
      value,
      metadata,
    });
  }

  /** Delete a secret. */
  async deleteSecret(key: string): Promise<void> {
    await this.request<void>('DELETE', `/api/v1/vault/secrets/${encodeURIComponent(key)}`);
  }

  /** Encrypt data using a transit key. */
  async transitEncrypt(keyName: string, plaintext: string): Promise<TransitEncryptResponse> {
    return this.request<TransitEncryptResponse>('POST', '/api/v1/vault/transit/encrypt', {
      key_name: keyName,
      plaintext,
    });
  }

  /** Decrypt data using a transit key. */
  async transitDecrypt(keyName: string, ciphertext: string): Promise<TransitDecryptResponse> {
    return this.request<TransitDecryptResponse>('POST', '/api/v1/vault/transit/decrypt', {
      key_name: keyName,
      ciphertext,
    });
  }

  /** Create a new transit encryption key. */
  async createTransitKey(name: string, type?: string, exportable?: boolean): Promise<TransitKey> {
    return this.request<TransitKey>('POST', '/api/v1/vault/transit/keys', {
      name,
      type,
      exportable,
    });
  }

  /** List all transit encryption keys. */
  async listTransitKeys(): Promise<TransitKey[]> {
    return this.request<TransitKey[]>('GET', '/api/v1/vault/transit/keys');
  }

  /** Get vault audit log. */
  async getAuditLog(): Promise<AuditLogResponse> {
    return this.request<AuditLogResponse>('GET', '/api/v1/vault/audit');
  }
}
